[2-C] 강태산 JPA 게시판 구현 미션 제출

Pawer0223 opened the pull request:
## API
- 게시글 조회
    - 페이징 조회 (GET "/posts")
    - 단건 조회 (GET "/posts/{id}")
- 게시글 작성 (POST "/posts")
- 게시글 수정 (PATCH "/posts")

## 문서화
- REST-DOCS

## PR 포인트 & 궁금한 점
- 가감없는 피드백 부탁드립니다 ! 달게 받겠습니다

Pawer0223 committed:
feat: User Entity 생성

- User Entity 생성

Pawer0223 committed:
refactor: CreationTimestamp -> CreatedDate 변경

- CreationTimestamp -> CreatedDate 변경

Pawer0223 committed:
refactor: 임베디드 타입 추가

- 임베디드 타입 추가

Pawer0223 committed:
feat: Post Entity 생성

- Post Entity 생성

Pawer0223 committed:
refactor: 공통 부분 분리, 테스트 분리

- 공통 부분 분리, 테스트 분리
- Embeddable 테스트 분리

Pawer0223 committed:
feat: 연관관계 설정

- 연관관계 설정

Pawer0223 committed:
refactor: Lombok 제거

- Lombok 제거

Pawer0223 committed:
chore: 디렉토리 이름 변경

- 디렉토리 이름 변경

Pawer0223 committed:
feat: PostService 구현

- PostService 구현
- Dto 정의
- Convertor 정의

Pawer0223 committed:
feat: PostController 구현

- PostController 구현

Pawer0223 committed:
feat: RestDocs 적용

- RestDocs 적용

Pawer0223 committed:
feat: update api 추가

- update api 추가

Pawer0223 committed:
chore: lombok 의존성 제거

- lombok 의존성 제거

sdardew commented on src/test/java/com/prgrms/boardapp/common/UserCreateUtil.java:
```diff
@@ -0,0 +1,53 @@
+
+    public static UserDto createUserDto() {
+        return UserDto.builder()
+                .id(userId++)
```
userId는 공유하는것 같은데 synchronized 처리가 필요하지 않을까요?

sdardew replied:
공용 변수를 사용하고 있는데 동기화 처리가 필요하지 않을까요?

sdardew commented on src/test/java/com/prgrms/boardapp/common/PostCreateUtil.java:
```diff
@@ -0,0 +1,47 @@
+
+    public static Post createPostWithId() {
+        return Post.builder()
+                .id(postId++)
```
postId도 공용 변수라서 동기화 처리가 필요한 것 같습니다!

sdardew commented on src/test/java/com/prgrms/boardapp/common/PostCreateUtil.java:
```diff
@@ -0,0 +1,47 @@
+
+    public static Post createPostWithId() {
+        return Post.builder()
+                .id(postId++)
```
공용 변수를 사용하고 있는데 synchronized를 사용해야 하지 않을까요??

ksy90101 replied:
이런경우에 synchronized를 사용하는게 맞을까요?
왜 synchronized를 사용했나요?

sdardew replied:
postId는 전역 변수이고, ++로 값을 변경하는 과정에서 여러 스레드가 변경할 가능성이 있기 때문에 thred-safe 하지 않다고 생각했습니다. 그래서 synchronized를 사용해야 한다고 생각했습니다.

Pawer0223 replied:
저도 피드백주신 부분이 문제될 수 있다고 생각했고.. synchronized만 붙이면 기존의 다른 코드는 수정을 안해도 되어서 그렇게 보완해보았습니다.

다른방법으로 @BeforeEach, @AfterEach 를 적용해볼까도 고민했는데.. 
오히려 Id가 중복되지 않게 테스트케이스가 동작할 수 있고, 모든 테스트에 해당 어노테이션을 붙여주지 않아도 되서 나쁘지않겠다..? 고 생각했습니다 !

ksy90101 replied:
테스트는 최대한 격리가 되어야 하는데요.
스레드세이프한걸 synchronized로 푸는건 안좋은거 같아요.

sdardew commented on src/main/java/com/prgrms/boardapp/constant/UserErrMsg.java:
```diff
@@ -0,0 +1,22 @@
+import static com.prgrms.boardapp.model.User.*;
+
+public enum UserErrMsg {
+    NAME_LENGTH_ERR_MSG("이름의 최대길이는 " + NAME_MAX_LENGTH +" 자 입니다."),
+
+    CREATED_LENGTH_ERR_MSG("취미의 최대길이는 " + CREATED_BY_MAX_LENGTH +" 자 입니다."),
```
최대 길이는 이미 정해져있는 값인 것 같은데, 문자열 연산보다는 바로 표기하는게 낫지 않나요?

Pawer0223 replied:
에러 메시지나 Validation 그리고 테스트 코드에서도 MAX 상수 값이 참조되고 있어서 변경에 따른 영향도를 줄이려고 문자열 연산을 선택했습니다 !

Pawer0223 committed:
refactor: readOnly 옵션 추가

- readOnly 옵션 추가

Pawer0223 committed:
refactor: 공유자원 상호배제

- 공유자원 상호배제

ksy90101 commented on src/main/java/com/prgrms/boardapp/controller/PostController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<Map<String, Long>> save(@RequestBody PostDto postDto) {
```
Dto라는 단어보단 Request, Response로 쓰는게 더 명시적이더라구요.

ksy90101 commented on src/main/java/com/prgrms/boardapp/controller/PostController.java:
```diff
@@ -0,0 +1,45 @@
+    @PostMapping
+    public ResponseEntity<Map<String, Long>> save(@RequestBody PostDto postDto) {
+        Long savedId = postService.save(postDto);
+        Map<String, Long> body = Map.of(SAVE_RESPONSE_KEY, savedId);
+        return ResponseEntity.ok(body);
```
생성시에서는 created를 많이 사용하는 편이에요.
REST API를 좀 더 찾아보실까요?

Pawer0223 replied:
지난번에도 같은 피드백 받았는데.. 부끄럽네요.
변경했습니다. 감사합니다 !
- [참조1](https://stackoverflow.com/questions/19199872/best-practice-for-restful-post-response)
- [참조2](https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#useful-post-responses)

ksy90101 commented on src/main/java/com/prgrms/boardapp/controller/PostController.java:
```diff
@@ -0,0 +1,45 @@
+        return ResponseEntity.ok(body);
+    }
+
+    @PatchMapping
```
이런 경우에는 Patch가 맞을까요? Put이 맞을까요?

Pawer0223 replied:
Post의 content, title만 변경이되고 있어 부분변경을 의미하는 Patch가 더 적절하다고 생각했습니다 !

ksy90101 replied:
그렇다면 모든 entity의 변경은 부분변경일까요?
어떤 경우가 전체 변경인가요?

Pawer0223 replied:
모든 Entity의 변경도 부분변경이 맞다고 이해하고 있습니다. 더불어 전체 변경은 해당 데이터가 대체되는 경우라 생각합니다.

따라서 API 요청데이터로 대체를 의미하는 경우에 Put이 적절하다고 생각됩니다 ! 위와 같은 경우에서 Put메서드 요청이라면, 나머지 필드는 null값이 되어야할것 같습니다.

- [참조1](https://tecoble.techcourse.co.kr/post/2020-08-17-put-vs-patch/)
- [참조2](https://papababo.tistory.com/entry/HTTP-METHOD-PUT-vs-PATCH-%EC%B0%A8%EC%9D%B4%EC%A0%90#rp)

ksy90101 commented on src/main/java/com/prgrms/boardapp/converter/Converter.java:
```diff
@@ -0,0 +1,60 @@
+        for (int i = 0; i < user.getPosts().size(); i++) {
+            postDtoList.add(this.convertPostDto(user.getPosts().get(i)));
+        }
+        return postDtoList;
```
stream으로 충분히 가능해 보이는데요.
한번 stream으로 해볼까요?

ksy90101 commented on src/main/java/com/prgrms/boardapp/converter/Converter.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.List;
+
+@Component
+public class Converter {
```
👍

ksy90101 commented on src/main/java/com/prgrms/boardapp/dto/PostDto.java:
```diff
@@ -0,0 +1,127 @@
+
+public class PostDto {
+    private Long id;
+    private UserDto userDto;
```
이렇게 하면 내려갈떄 key값이 어떻게 될까요?

Pawer0223 replied:
``` json
{ 
   "id" : "1" , 
  "userDto" : ""
   ...
}
```

와 같이 전달되어 Dto를 불필요하게 노출시키게 됩니다. user로 변경하였습니다.
피드백 감사합니다 !

ksy90101 commented on src/main/java/com/prgrms/boardapp/dto/PostDto.java:
```diff
@@ -0,0 +1,127 @@
+        return userDto;
+    }
+
+    public void setUserDto(UserDto userDto) {
+        this.userDto = userDto;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
+
+    public String getContent() {
+        return content;
+    }
+
+    public void setContent(String content) {
+        this.content = content;
+    }
```
setter는 필요 없습니다.
한번 Spring이 json을 어떻게 직렬화 하는지 알아볼까요?

Pawer0223 replied:
Spring에서는 Jackson 라이브러리를 통해 직렬화/역직렬화가 수행되는데, 내부적으로 getter를 통해 처리하고 있어 setter가 불필요 하다 이해했습니다.

- [참조](https://www.baeldung.com/jackson-field-serializable-deserializable-or-not#a-public-field)
- setter는 `A Setter Makes a Non-Public Field Deserializable Only` 역할로 사용된다고 설명하고 있네요.. `private 필드의 역직렬화만 가능하게 만든다 ?` (요건 아직 정확히 이해가 되지는 않습니다..ㅜㅜ)

추가적으로 Get요청을 처리할 때는 Jackson라이브러리를 사용하지 않고 Java Bean 방식을 사용하기 때문에 Setter가 필요합니다. 현재 Dto가 Get 요청에서는 활용되고 있지 않는데.. 만약 사용된다면 다시 setter를 정의하거나, setter를 사용하지 않고 직렬화/역직렬화 하려면 initDirectFieldAccess 옵션을 주어야한다고 이해했습니다.

- [참조](https://jojoldu.tistory.com/407)

dto에서 setter를 무분별하게 사용했는데 좋은 공부가 되었습니다. 피드백 감사합니다 !

ksy90101 commented on src/main/java/com/prgrms/boardapp/model/CommonEmbeddable.java:
```diff
@@ -0,0 +1,80 @@
+public class CommonEmbeddable {
+    @Column(
+            length = 50
+    )
```
굳이 이렇게 적어야 할까요?
개행이 가끔은 가독성을 떨어트리긴 합니다.

ksy90101 commented on src/main/java/com/prgrms/boardapp/model/CommonEmbeddable.java:
```diff
@@ -0,0 +1,80 @@
+    )
+    private LocalDateTime createdAt;
+
+    public static final int CREATED_BY_MAX_LENGTH = 50;
```
필드보다 상수가 더 위에.
아울러 private이 좋습니다 :)

Pawer0223 replied:
위치 변경했습니다 ! public은 해당 상수를 에러메시지 출력과정에서도 참조하고 있기때문에 public으로 선언했습니다.
피드백 감사합니다 !

ksy90101 commented on src/main/java/com/prgrms/boardapp/model/Post.java:
```diff
@@ -0,0 +1,133 @@
+        user.getPosts().add(this);
+    }
+
+    public void changeTitle(String title) {
+        this.title = title;
+    }
+
+    public void changeContent(String content) {
+        this.content = content;
+    }
```
따로 변경될 일이 있을까요?
같이 변경될 확률이 훨씬 높을꺼 같은데요 :)

Pawer0223 replied:
현재 어플리케이션에서는 같이 변경되는 경우만 존재합니다.
피드백 주신대로 수정하는게 더 좋을것 같아요. 감사합니다 !

ksy90101 commented on src/test/java/com/prgrms/boardapp/common/EntityManagerTest.java:
```diff
@@ -0,0 +1,24 @@
+import javax.persistence.EntityTransaction;
+
+@DataJpaTest
+public class EntityManagerTest {
```
이 미션에서까지 이런 테스트가 필요할까요?

Pawer0223 replied:
연관관계 편의 메서드를 함께 정의했는데.. 아직 연관관계 메서드가 호출되는 Service가 존재하지 않아 최소한의 검증을 하고자 구현해보았습니다.. !

ksy90101 reviewed (changes requested):
태산님 미션하시느라 너무 고생 많으셨습니다 💯 
간단한 피드백을 남겼으니 확인해주세요~

Pawer0223 committed:
chore: 변수이름 변경

- 변수이름 변경

Pawer0223 committed:
refactor: 상수 위치변경, 개행 제거

- 상수 위치변경, 개행 제거

Pawer0223 committed:
refactor: 생성은 201응답 with Location

- 생성은 201응답 with Location

hyuk0309 commented on src/main/java/com/prgrms/boardapp/converter/Converter.java:
```diff
@@ -0,0 +1,60 @@
+package com.prgrms.boardapp.converter;
```
제 생각에 현제 Converter는 Service Layer에서만 사용되는 것 같아요! 
Converter 클래스를 Service와 같은 패키지에 두고, 접근 제어자를 private-package로 설정하는 것도 하나의 방법일 것 같아요!

[이펙티브자바 아이템15](https://www.notion.so/15-f2949b0690174fad8a7df876d8f2bbc6)
제가 책 읽고 정리한 링크입니다. 혹시 몰라서 링크 남겼어요! 😄

Pawer0223 replied:
이런 고민은 한번도 안해봤는데.. 좋은 고민이네요.. 공유주신글 감사히 잘 읽었습니다 👍

hyuk0309 commented on src/main/java/com/prgrms/boardapp/dto/PostDto.java:
```diff
@@ -0,0 +1,127 @@
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+public class PostDto {
```
게시글 생성, 수정, 조회 모두 같은 Dto를 사용하신 이유가 있으신가요? 🤔

Pawer0223 replied:
음.. 비슷한 필드가 주로 사용되는데 Dto를 여러개로 나누는것에 비효율을 느꼈습니다.
또한 Dto를 사용한 주 목적은 Entity를 노출하지 않기 위함이었습니다. 그래서 크게 필요성을 못느끼기도 했어요 !

hyuk0309 commented on src/main/java/com/prgrms/boardapp/dto/PostDto.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Override
+    public String toString() {
```
SpringSecurity 강의에서 강사님은 ToStringBuilder를 사용하시는 것 같습니다! 
정답은 없지만, 저도 StringBuilder를 사용하는 게 성능 면에서 더 유리하다고 생각합니다! 🙇‍♂️

Pawer0223 replied:
오 ToStringBuilder를 몰랐습니다.. 확실히 가독성이나 유지보수 측면에서 더 장점이 있어보이네요.
피드백 감사합니다 !

hyuk0309 commented on src/main/java/com/prgrms/boardapp/dto/PostDto.java:
```diff
@@ -0,0 +1,127 @@
+                '}';
+    }
+
+    public static PostDtoBuilder builder() {
```
👍
빌더 패턴을 활용하니까 converter 가독성이 확실히 좋은 것 같아요! (태산님 코드 참고해 저도 도입해 봐야겠습니다! 👨‍💻)

hyuk0309 commented on src/main/java/com/prgrms/boardapp/exception/ApiControllerAdvice.java:
```diff
@@ -0,0 +1,26 @@
+import javax.persistence.EntityNotFoundException;
+
+@ControllerAdvice
+public class ApiControllerAdvice {
```
👍

hyuk0309 commented on src/main/java/com/prgrms/boardapp/model/Post.java:
```diff
@@ -0,0 +1,133 @@
+    protected Post() {
+    }
+
+    public Post(Long id, String title, String content, CommonEmbeddable commonEmbeddable, User user) {
```
빌더 패턴을 통해 객체를 생성할 수 있으면, 생성자는 클라이언트가 사용하지 못하게 막는 게 좋지 않을까요? 🤔

Pawer0223 replied:
전혀 고려하지 못한부분이네요.. 실제로 private으로 해도 현재 테스트케이스는 모두 이상이없네요.
피드백 감사합니다 !

hyuk0309 commented on src/main/java/com/prgrms/boardapp/service/PostService.java:
```diff
@@ -0,0 +1,55 @@
+    private final PostRepository postRepository;
+    private final Converter converter;
+
+    private static final String NOT_FOUND_POST_ERR_MSG = "게시글 정보를 찾을 수 없습니다.";
```
PostErrMsg.java에서 같이 관리해 주는 게 좋지 않을까요? 🤔

Pawer0223 replied:
오, 그렇게 분류해도 이상하지 않을것 같네요.. 꼼꼼한 리뷰 감사함니다 ~

hyuk0309 reviewed (commented):
안녕하세요! 😀

부족하지만 개인적인 의견 남겨봤습니다! 
Builder 패턴과 Api 예외처리 배우고 갑니다! 🙇‍♂️

고생하셨어요! 👍

Pawer0223 committed:
refactor: 에러메시지 공통화

- 에러메시지 공통화

Pawer0223 committed:
refactor: synchronized 다시 제거

- 테스트 과정에서 Synchronized가 사용되면 격리성이 깨진다.
- 정말 동시성이 깨지는가? -> 그렇다면 다른방법을
취하자.(BeforeEach, AfterEach)

Pawer0223 committed:
refactor: package-priavte 적용

- package-priavte 적용
- 이게 맞는가 ?

Pawer0223 committed:
refactor: setter제거 및 생성자 public->private 변경

- setter제거 및 생성자 public->private 변경

Pawer0223 committed:
refactor: Dto, Converter 변경

- Dto 분리
- Converter 간략화. 사용되지 않는 필드는 제거
- dto를 key로 가지는 응답변경

Pawer0223 committed:
refactor: Service findById 테스트 추가

- Service findById 테스트 추가
